src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,176 @@
+package is.hail.io.index
+
+import is.hail.annotations.{Annotation, Region, RegionValueBuilder}
+import is.hail.expr.types._
+import is.hail.io.CodecSpec
+import is.hail.utils._
+import is.hail.utils.richUtils.ByteTrackingOutputStream
+import org.apache.hadoop.conf.Configuration
+import org.json4s.Formats
+import org.json4s.jackson.Serialization
+
+case class IndexMetadata(
+  fileVersion: Int,
+  branchingFactor: Int,
+  height: Int,
+  keyType: String,
+  nKeys: Long,
+  indexPath: String,
+  rootOffset: Long,
+  attributes: Map[String, Any]
+)
+
+case class IndexNodeInfo(
+  fileOffset: Long,
+  firstIndex: Long,
+  firstKey: Annotation,
+  firstKeyOffset: Long
+)
+
+class IndexWriter(
+  hConf: Configuration,
+  path: String,
+  keyType: Type,
+  branchingFactor: Int = 1024,
+  attributes: Map[String, Any] = Map.empty[String, Any]) extends AutoCloseable {
+
+  private val indexFile = path + "/index"
+  private val metadataFile = path + "/metadata.json.gz"
+
+  private var elementIdx = 0L
+  private var rootOffset = 0L
+
+  private val rvb = new RegionValueBuilder()
+  private val region = new Region()
+  rvb.set(region)
+
+  private val leafNode = new LeafNodeBuilder(keyType)
+  private val internalNodes = new ArrayBuilder[InternalNodeBuilder]()
+
+  private val trackedOS = new ByteTrackingOutputStream(hConf.unsafeWriter(indexFile))
+  private val codecSpec = CodecSpec.default
+  private val leafEncoder = codecSpec.buildEncoder(leafNode.typ)(trackedOS)
+  private val internalEncoder = codecSpec.buildEncoder(InternalNodeBuilder.typ(keyType))(trackedOS)
+
+  private def calcDepth: Int =
+    math.max(1, (math.log10(elementIdx) / math.log10(branchingFactor)).ceil.toInt) // max necessary for array of length 1 becomes depth = 0
+
+  private def height: Int = math.max(calcDepth - 1, 1) + 1 // ensure always at least one internal level and one leaf level
```

Comment:
The last internal node is one that is setup but never written. Otherwise, you don't have access to the first child by the time you're ready to add an extra level. So the rule is actually `height = internalNodes.length`.